File: py/src/tradekit/core/money.py
"""Monetary amounts as integer minor units.

A ``Money`` value is a count of the currency's smallest unit -- paise for INR,
cents for USD. Integers are used rather than ``float`` because accumulated
profit and loss must be exact, and rather than :class:`decimal.Decimal` because
the same representation has to exist unchanged in Go, in Python and in a SQLite
``INTEGER`` column.

This module mirrors ``go/core/money`` exactly. Every function here has a
counterpart there with the same name in Go's casing, and both round half away
from zero.
"""

from __future__ import annotations

import math
import re
from decimal import ROUND_HALF_UP, Decimal
from typing import NewType

__all__ = [
    "ZERO",
    "Money",
    "ceil_to_tick",
    "floor_to_tick",
    "format",
    "from_major",
    "major",
    "mul",
    "mul_fraction",
    "parse",
    "round_to_tick",
]

Money = NewType("Money", int)
"""An amount in the currency's minor unit."""

ZERO: Money = Money(0)

_SCALE = 100
"""Minor units in one major unit. Both currencies tradekit targets use 100."""

_AMOUNT = re.compile(r"^[+-]?(?:\d+(?:\.\d{0,2})?|\.\d{1,2})$")


def parse(s: str) -> Money:
    """Read a decimal string such as ``"1234.56"`` into minor units.

    The conversion never passes through binary floating point, so a price that
    a broker sent as text arrives exactly.

    Args:
        s: A decimal amount with at most two fractional digits.

    Returns:
        The amount in minor units.

    Raises:
        ValueError: If ``s`` is not a decimal amount, or carries more than two
            decimal places. More precision than the exchange can quote is a bug
            upstream, and rounding it away here would hide it.

    """
    text = s.strip()
    if not _AMOUNT.match(text):
        raise ValueError(f"money: cannot parse amount: {s!r}")
    return Money(int(Decimal(text).scaleb(2).to_integral_value(rounding=ROUND_HALF_UP)))


def format(m: Money) -> str:  # noqa: A001 - mirrors Go's Money.String
    """Render an amount with exactly two decimals, so it round-trips ``parse``."""
    sign = "-" if m < 0 else ""
    v = abs(int(m))
    return f"{sign}{v // _SCALE}.{v % _SCALE:02d}"


def from_major(v: float) -> Money:
    """Convert a whole-currency amount, rounding half away from zero.

    Intended for configuration and fixtures. Prices from a broker API should go
    through :func:`parse` instead, which avoids floating point entirely.
    """
    return Money(_round_half_away(v * _SCALE))


def major(m: Money) -> float:
    """Return the amount as a whole-currency float, for display and plotting.

    Never round-trip through this: ``from_major(major(m))`` is not guaranteed to
    equal ``m`` for very large amounts.
    """
    return int(m) / _SCALE


def mul(m: Money, qty: int) -> Money:
    """Scale an amount by a whole quantity, as in price times shares."""
    return Money(int(m) * int(qty))


def mul_fraction(m: Money, f: float) -> Money:
    """Scale an amount by a ratio, rounding half away from zero.

    Used for charge rates, percentages and ATR multiples.
    """
    return Money(_round_half_away(int(m) * f))


def round_to_tick(m: Money, tick: Money) -> Money:
    """Return the nearest multiple of ``tick``.

    A non-positive tick returns the amount unchanged: refusing to round is safer
    than inventing a tick size the venue may not use.
    """
    if tick <= 0:
        return m
    return Money(_round_quotient(int(m), int(tick)) * int(tick))


def floor_to_tick(m: Money, tick: Money) -> Money:
    """Return the largest multiple of ``tick`` not greater than the amount.

    Use it for a sell limit, where rounding up could place the order outside
    the book.
    """
    if tick <= 0:
        return m
    # Python's // already floors for negatives; using float division here
    # would lose precision on large amounts.
    return Money((int(m) // int(tick)) * int(tick))


def ceil_to_tick(m: Money, tick: Money) -> Money:
    """Return the smallest multiple of ``tick`` not less than the amount.

    Use it for a buy limit, for the mirror-image reason to :func:`floor_to_tick`.
    """
    if tick <= 0:
        return m
    return Money(-((-int(m)) // int(tick)) * int(tick))


def _round_half_away(v: float) -> int:
    """Round half away from zero.

    Python's built-in ``round`` uses banker's rounding, which would disagree
    with the Go implementation on every exact half. This is the single rounding
    rule tradekit uses.
    """
    if math.isnan(v) or math.isinf(v):
        return 0
    return math.floor(v + 0.5) if v >= 0 else math.ceil(v - 0.5)


def _round_quotient(a: int, b: int) -> int:
    """Divide and round half away from zero, without floating point.

    Computed on magnitudes and re-signed at the end. Working with Python's
    flooring ``divmod`` directly gets the negative cases wrong, because
    "half away from zero" is defined against truncation, not against the floor.
    """
    magnitude, divisor = abs(a), abs(b)
    q, r = divmod(magnitude, divisor)
    if 2 * r >= divisor:
        q += 1
    return q if (a < 0) == (b < 0) else -q

File: go/core/money/money.go
// Package money represents monetary amounts as integer minor units.
//
// A Money value is a count of the currency's smallest unit — paise for INR,
// cents for USD. Integers are used rather than float64 because accumulated
// profit and loss must be exact, and rather than a decimal library because the
// same representation has to exist unchanged in Go, in Python and in a SQLite
// INTEGER column.
//
// A Money value carries no currency tag. Mixing currencies in one calculation
// is prevented by construction elsewhere: a Trade, a Position and an Account
// each belong to a single instrument or venue.
package money

import (
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"
)

// Money is an amount in the currency's minor unit.
type Money int64

// Zero is the additive identity, provided for readability at call sites.
const Zero Money = 0

// scale is the number of minor units in one major unit. Both currencies
// tradekit targets (INR, USD) use 100, and a currency that does not would need
// its own type rather than a variable here.
const scale = 100

// ErrParse reports a string that is not a valid decimal amount.
var ErrParse = errors.New("money: cannot parse amount")

// FromMajor converts a whole-currency amount, rounding half away from zero.
//
// It is intended for configuration and test fixtures, where the input is a
// literal such as 1234.56. Prices arriving from a broker API should go through
// Parse instead, which does not pass through binary floating point at all.
func FromMajor(v float64) Money {
	return Money(roundHalfAwayFromZero(v * scale))
}

// Major returns the amount as a whole-currency float. It is for display and
// for handing values to code that cannot accept integers, such as a plotting
// library. Never round-trip through it: FromMajor(m.Major()) is not guaranteed
// to equal m for very large amounts.
func (m Money) Major() float64 { return float64(m) / scale }

// Parse reads a decimal string such as "1234.56", "-0.05" or "1234" into
// minor units, without floating-point intermediate steps.
//
// More than two decimal places is an error rather than a silent truncation: a
// price the exchange cannot quote is a bug upstream, and rounding it here
// would hide it.
func Parse(s string) (Money, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0, fmt.Errorf("%w: empty string", ErrParse)
	}

	neg := false
	switch s[0] {
	case '-':
		neg, s = true, s[1:]
	case '+':
		s = s[1:]
	}

	intPart, fracPart, hasFrac := strings.Cut(s, ".")
	if intPart == "" && !hasFrac {
		return 0, fmt.Errorf("%w: %q", ErrParse, s)
	}
	if intPart == "" {
		intPart = "0"
	}
	if len(fracPart) > 2 {
		return 0, fmt.Errorf("%w: %q has more than 2 decimal places", ErrParse, s)
	}
	// strconv.ParseInt would accept a second sign here ("--1") and any
	// underscore separators Go literals allow, neither of which is a price.
	if !allDigits(intPart) || !allDigits(fracPart) {
		return 0, fmt.Errorf("%w: %q", ErrParse, s)
	}

	units, err := strconv.ParseInt(intPart, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("%w: %q", ErrParse, s)
	}

	var minor int64
	if hasFrac && fracPart != "" {
		// "5" means 50 paise, "05" means 5 paise.
		padded := fracPart + strings.Repeat("0", 2-len(fracPart))
		minor, err = strconv.ParseInt(padded, 10, 64)
		if err != nil {
			return 0, fmt.Errorf("%w: %q", ErrParse, s)
		}
	}

	total := units*scale + minor
	if neg {
		total = -total
	}
	return Money(total), nil
}

// MustParse is Parse for constants and test fixtures, panicking on bad input.
func MustParse(s string) Money {
	m, err := Parse(s)
	if err != nil {
		panic(err)
	}
	return m
}

// String renders the amount with exactly two decimal places and no thousands
// separators, so that it round-trips through Parse.
func (m Money) String() string {
	neg := m < 0
	v := int64(m)
	if neg {
		v = -v
	}
	s := fmt.Sprintf("%d.%02d", v/scale, v%scale)
	if neg {
		return "-" + s
	}
	return s
}

// Mul scales the amount by a whole quantity, as in price times shares.
func (m Money) Mul(qty int64) Money { return Money(int64(m) * qty) }

// MulFraction scales the amount by a ratio — a charge rate, a percentage, an
// ATR multiple — rounding half away from zero.
func (m Money) MulFraction(f float64) Money {
	return Money(roundHalfAwayFromZero(float64(m) * f))
}

// Abs returns the magnitude of the amount.
func (m Money) Abs() Money {
	if m < 0 {
		return -m
	}
	return m
}

// Sign reports -1, 0 or +1.
func (m Money) Sign() int {
	switch {
	case m < 0:
		return -1
	case m > 0:
		return 1
	default:
		return 0
	}
}

// RoundToTick returns the nearest multiple of tick, which is how a limit price
// must be expressed for the exchange to accept it. A non-positive tick returns
// the amount unchanged, because refusing to round is safer than inventing a
// tick size the venue may not use.
func (m Money) RoundToTick(tick Money) Money {
	if tick <= 0 {
		return m
	}
	return roundQuotient(int64(m), int64(tick)) * tick
}

// FloorToTick returns the largest multiple of tick not greater than the
// amount. Use it for a sell limit, where rounding up could place the order
// outside the book.
func (m Money) FloorToTick(tick Money) Money {
	if tick <= 0 {
		return m
	}
	q := int64(m) / int64(tick)
	if int64(m)%int64(tick) < 0 {
		q--
	}
	return Money(q) * tick
}

// CeilToTick returns the smallest multiple of tick not less than the amount.
// Use it for a buy limit, for the mirror-image reason to FloorToTick.
func (m Money) CeilToTick(tick Money) Money {
	if tick <= 0 {
		return m
	}
	q := int64(m) / int64(tick)
	if int64(m)%int64(tick) > 0 {
		q++
	}
	return Money(q) * tick
}

// allDigits reports whether s consists only of ASCII digits. An empty string
// is accepted: it is the absent fractional part of "12.".
func allDigits(s string) bool {
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return false
		}
	}
	return true
}

// roundQuotient divides and rounds half away from zero, without floating point.
func roundQuotient(a, b int64) Money {
	q, r := a/b, a%b
	if r == 0 {
		return Money(q)
	}
	twice := 2 * r
	if twice < 0 {
		twice = -twice
	}
	if twice >= b {
		if (a < 0) != (b < 0) {
			q--
		} else {
			q++
		}
	}
	return Money(q)
}

// roundHalfAwayFromZero matches Python's decimal ROUND_HALF_UP and is the
// single rounding rule tradekit uses. Go's math.Round already rounds half away
// from zero; this wrapper exists so the rule is named once and the Python
// implementation has an obvious counterpart.
func roundHalfAwayFromZero(v float64) int64 {
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return 0
	}
	return int64(math.Round(v))
}
